modules/artifact/src/main/java/org/mule/runtime/module/artifact/api/classloader/MuleArtifactClassLoader.java

```diff
@@ -307,47 +309,39 @@ public void dispose() {
     });
     descriptorMapping.clear();
 
-    try {
-      clearReferences();
-    } catch (Throwable t) {
-      reportPossibleLeak(t, artifactId);
-    }
-
-    try {
-      if (shouldReleaseJdbcReferences) {
-        createResourceReleaserInstance().release();
-      }
-    } catch (Throwable t) {
-      reportPossibleLeak(t, artifactId);
-    }
-
-    try {
-      if (shouldReleaseIbmMQResources) {
-        new IBMMQResourceReleaser(this).release();
-      }
-    } catch (Throwable t) {
-      reportPossibleLeak(t, artifactId);
+    // When running on Java versions greater than 11, the resource releaser logic from the Mule Runtime will not be used.
+    // The resource releasing responsibility will be delegated to each extension instead.
+    if (IS_JAVA_VERSION_AT_MOST_11) {
+      addLegacyExtensionsResourceReleasers();
     }
 
-    try {
-      if (shouldReleaseActiveMQReferences) {
-        new ActiveMQResourceReleaser(this).release();
-      }
-    } catch (Throwable t) {
-      reportPossibleLeak(t, artifactId);
-    }
+    resourceReleaserExecutor.executeResourceReleasers();
 
     super.dispose();
     shutdownListeners();
   }
 
-  private void clearReferences() {
-    classLoaderReferenceReleaser.release();
-    scalaClassValueReleaserInstance.release();
-    mvelClassLoaderReleaserInstance.release();
+  @Deprecated
+  private void addLegacyExtensionsResourceReleasers() {
+    resourceReleaserExecutor.addResourceReleaser(() -> new AwsIdleConnectionReaperResourceReleaser(this));
+    resourceReleaserExecutor.addResourceReleaser(ScalaClassValueReleaser::new);
+
+    if (shouldReleaseJdbcReferences) {
+      resourceReleaserExecutor.addResourceReleaser(this::createResourceReleaserInstance);
+    }
+    if (shouldReleaseIbmMQResources) {
+      resourceReleaserExecutor.addResourceReleaser(() -> new IBMMQResourceReleaser(this));
+    }
+    if (shouldReleaseActiveMQReferences) {
+      resourceReleaserExecutor.addResourceReleaser(() -> new ActiveMQResourceReleaser(this));
+    }
+  }
+
+  private void reportPossibleLeak(Throwable t) {
+    reportPossibleLeak(t, artifactId);
   }
 
-  void reportPossibleLeak(Throwable t, String artifactId) {
+  protected void reportPossibleLeak(Throwable t, String artifactId) {
```

Comment:
this method makes no sense:
* The message always gets logged only that the level gets downgraded if in debug mode. (???????)
* The point of asking `isDebugEnabled` is not instantiating the message, which always happens anyway.

Reply:
done, changed to always ERROR level and fixed the stack trace logging issue.